### PR TITLE
[captive_portal] Add DNS server support for ESP-IDF framework

### PR DESCRIPTION
**Pull request by bdraco**
# What does this implement/fix?

This PR adds DNS server functionality to the captive portal component for ESP-IDF framework builds. Previously, the captive portal only worked with Arduino framework on ESP32 because the DNS server was not implemented for ESP-IDF.

### Event Loop Integration
The DNS server integrates with ESPHome's existing event loop infrastructure:
- **Consistency**: Matches the Arduino implementation pattern (processNextRequest in loop)
- **Integration**: Uses ESPHome's loop-monitored socket for immediate wake-on-data
- **Simplicity**: Single-threaded operation, no synchronization needed
- **Proven pattern**: Follows same approach as other UDP components in ESPHome

### Single Buffer Architecture
The implementation uses a single buffer for both request and response (modifying in-place):
- **Memory efficient**: One 192-byte buffer instead of two
- **Safe**: Buffer size check ensures answer fits (response = request + 12 byte answer)
- **Performance**: Eliminates memory copying between buffers
- **Validated**: Comprehensive testing confirms no buffer overflows possible

### Key Changes to Captive Portal Behavior
1. **Cleaner design**: Removed onNotFound handler - now simply `canHandle()` returns true for all GET requests while portal is active, false when inactive (no interference with normal web server)
2. **Simplified request handling**: Single code path serves portal page for any URL (only when active)
3. **Fixed OS detection**: Responding to all URLs ensures detection probes get unexpected response, triggering portal UI
4. **Loop lifecycle management**: Loop disabled in setup, enabled only when portal is active, disabled when stopped

The implementation:
- Adds a lightweight DNS server that runs in the ESPHome event loop (no separate task needed)
- Responds to all DNS A record queries with the ESP's IP address to trigger captive portal detection
- Properly handles all OS-specific captive portal detection endpoints
- Uses non-blocking sockets with ESPHome's loop-monitored socket infrastructure
- Includes comprehensive buffer overflow protection and malformed packet handling

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

- Fixes captive portal not working on ESP-IDF framework builds
- Enables proper captive portal detection on iPhone/iOS devices

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**

- Not required (no user-facing changes, captive portal already documented)

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [x] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
# Example config.yaml - no changes needed
wifi:
  ap:
    ssid: "MyESPHome"
    password: "12345678"

captive_portal:
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

---

## Key Improvements

1. **DNS Server Implementation for ESP-IDF**:
   - Lightweight implementation using ESPHome's event loop (no FreeRTOS task needed)
   - Processes one DNS packet per loop iteration to avoid blocking
   - Uses non-blocking sockets with MSG_DONTWAIT flag
   - Comprehensive input validation and buffer overflow protection

2. **Captive Portal Detection**:
   - Handles all GET requests when active (triggers OS detection)
   - Properly responds to OS-specific endpoints (Apple's /hotspot-detect.html, Android's /generate_204, etc.)
   - Ensures reliable portal detection across different operating systems

3. **Memory and Performance Optimizations**:
   - Single buffer for request/response (saves memory)
   - Constexpr for all DNS constants (compile-time evaluation)
   - Loop enable/disable based on captive portal state
   - Memory only allocated while captive portal is active

4. **Safety and Reliability**:
   - Comprehensive buffer overflow protection
   - Handles malformed DNS packets gracefully
   - Filters non-A record queries (AAAA, HTTPS, SVCB, etc.)
   - Proper cleanup on shutdown

### Testing:

The implementation has been tested with:
- iPhone/iOS devices (captive portal popup appears immediately)
- Android devices (captive portal detection works)
- Various malformed packet scenarios (all handled safely)
- Buffer overflow attempts (all blocked)
- Extended runtime stability (no watchdog triggers)

DNS packet safety verified with comprehensive test suite covering:
- Valid queries of various sizes
- Buffer overflow attempts
- Malformed packets
- Truncated data
- Invalid query types
- Boundary conditions

### Performance Impact:

- **RAM**: 0 bytes (no static allocation - memory only used when captive portal is active)
- **Flash**: 236 bytes (0.01% increase)
- **CPU**: Negligible (only processes when DNS queries received)
- **Response Time**: Immediate (loop-monitored socket wakes the event loop when data arrives)

### CODE REVIEW

File: esphome/components/captive_portal/captive_portal.cpp [hunk 1]
```diff
@@ -45,8 +45,8 @@ void CaptivePortal::handle_config(AsyncWebServerRequest *request) {
   request->send(stream);
 }
 void CaptivePortal::handle_wifisave(AsyncWebServerRequest *request) {
-  std::string ssid = request->arg("ssid").c_str();
-  std::string psk = request->arg("psk").c_str();
+  std::string ssid = request->arg("ssid").c_str();  // NOLINT(readability-redundant-string-cstr)
+  std::string psk = request->arg("psk").c_str();    // NOLINT(readability-redundant-string-cstr)
```
Comment (bdraco) [inline esphome/components/captive_portal/captive_portal.cpp:49]: existing issue that linter started complaining about. We have to do this for Arduino String

[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/captive_portal/captive_portal.cpp:49]: No its arduino

File: esphome/components/captive_portal/dns_server_esp32_idf.cpp [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+  // Pointer to name in question (offset from start of packet)
+  answer->ptr_offset = htons(0xC000 | sizeof(DNSHeader));
+  answer->type = htons(DNS_QTYPE_A);
+  answer->dns_class = htons(DNS_QCLASS_IN);
+  answer->ttl = htonl(DNS_ANSWER_TTL);
+  answer->addr_len = htons(4);
+
+  // Get the raw IP address
+  ip4_addr_t addr = this->server_ip_;
```
[the remark this answers is not in the document]
Comment (bdraco) [reply esphome/components/captive_portal/dns_server_esp32_idf.cpp:188]: captive portal will always be ipv4

**Comment by bdraco** [comment]
Was verified ok on windows as well

https://discord.com/channels/429907082951524364/524177279270780928/1421851651920232449

**Review by jesserockz — APPROVED** [review]

**Comment by jesserockz** [comment]
Thanks @bdraco

**Comment by bdraco** [comment]
Thanks
